File: practice-projects/gin-ddd/internal/domain/event/event_publisher.go
package event

import "context"

// EventPublisher 事件发布器接口
type EventPublisher interface {
	// Publish 发布事件
	Publish(ctx context.Context, topic string, event DomainEvent) error

	// Close 关闭发布器
	Close() error
}

// EventConsumer 事件消费者接口
type EventConsumer interface {
	// Subscribe 订阅主题
	Subscribe(topic string, handler EventHandler) error

	// Start 启动消费者
	Start() error

	// Close 关闭消费者
	Close() error
}

// EventHandler 事件处理器
type EventHandler func(ctx context.Context, event DomainEvent) error

File: practice-projects/nestjs-ddd/src/domain/shared/events/event-publisher.ts
import { DomainEvent } from './domain-event';

export const EVENT_PUBLISHER = Symbol('EVENT_PUBLISHER');

/** 事件发布器接口（由 infrastructure 层实现） */
export interface EventPublisher {
  publish<T extends DomainEvent>(event: T): Promise<void>;
  publishAll<T extends DomainEvent>(events: T[]): Promise<void>;
}
